pkg/urbit/vere/io/ames.c

```diff
@@ -445,39 +686,164 @@ _ames_recv_cb(uv_udp_t*        wax_u,
               unsigned         flg_i)
 {
   u3_ames* sam_u = wax_u->data;
+  c3_o     pas_o = c3y;
+  c3_y*    byt_y = (c3_y*)buf_u->base;
+  c3_y*    bod_y = byt_y + 4;
+  u3_head  hed_u;
+
+  //  ensure a sane message size
+  //
+  if (4 >= nrd_i) {
+    pas_o = c3n;
+  }
+  //  unpack the packet header
+  //
+  else {
+    c3_w hed_w = (byt_y[0] <<  0)
+               | (byt_y[1] <<  8)
+               | (byt_y[2] << 16)
+               | (byt_y[3] << 24);
+
+    hed_u.ver_y = hed_w & 0x7;
+    hed_u.mug_l = (hed_w >> 3) & ((1 << 20) - 1);
+    hed_u.sac_y = (hed_w >> 23) & 0x3;
+    hed_u.rac_y = (hed_w >> 25) & 0x3;
+    hed_u.enc_o = (hed_w >> 27) & 0x1;
+  }
 
-  //  data present, and protocol version in header matches 0
+  //  ensure the protocol version matches ours
   //
-  //    XX inflexible, scry version out of ames
+  if ( c3y == pas_o
+    && (c3y == sam_u->fit_o)
+    && (sam_u->ver_y != hed_u.ver_y) )
+  {
+    pas_o = c3n;
+
+    sam_u->vet_d++;
+    if ( 0 == (sam_u->vet_d % 100) ) {
+      u3l_log("ames: %" PRIu64 " dropped for version mismatch\n", sam_u->vet_d);
+    }
+  }
+
+  //  ensure the mug is valid
+  //
+  if ( c3y == pas_o
+    && (hed_u.mug_l != _ca_mug_body(nrd_i - 4, bod_y)) )
+  {
+    pas_o = c3n;
+
+    sam_u->mut_d++;
+    if ( 0 == (sam_u->mut_d % 100) ) {
+      u3l_log("ames: %" PRIu64 " dropped for invalid mug\n", sam_u->mut_d);
+    }
+  }
+
+  //  unpack the body
+  //
+  c3_y  sen_y = 2 << hed_u.sac_y;
+  c3_y  rec_y = 2 << hed_u.rac_y;
+  c3_d  sen_d[2];
+  c3_d  rec_d[2];
+  c3_w  con_w = nrd_i - 4 - sen_y - rec_y;
+  c3_y* con_y = NULL;
+  if (c3y == pas_o) {
+    u3_noun sen = u3i_bytes(sen_y, bod_y);
+    u3_noun rec = u3i_bytes(rec_y, bod_y + sen_y);
+    u3r_chubs(0, 2, rec_d, rec);
+    u3r_chubs(0, 2, sen_d, sen);
+    u3z(sen); u3z(rec);
+
+    con_y = c3_malloc(con_w);
+    memcpy(con_y, bod_y + sen_y + rec_y, con_w);
+
+    //  ensure the content is cue-able
+    //
+    u3_noun con = u3i_bytes(con_w, con_y);
+    u3_noun pro = u3m_soft(0, u3qe_cue, con);
```

Comment:
I don't think it actually matters (due to the road transition), but this should be `u3ke_cue()`, to re-emphasize that `con` is being transferred. (It might also be better to just pass the `u3i_bytes()` return to further emphasize that, and avoid the intermediate variable.)